Automattic/Automattic-Tracks-iOS
# Add UIApplication.State tag for Submitted Sentry Events

Refs: p91TBi-2R9-p2#comment-928 and https://github.com/woocommerce/woocommerce-ios/issues/2371. 

This adds an `app.state` tag for events submitted to Sentry. Example:

<img width="1301" alt="this" src="https://user-images.githubusercontent.com/198826/83559175-5d614f80-a4d1-11ea-88c7-c6b8e83a504b.png">

This currently only supports iOS as I could not find a corresponding property for `NSApplication`. 

## Why? 

We're going to use this to retrieve more information for the crashes:

- https://github.com/woocommerce/woocommerce-ios/issues/2371
- https://github.com/woocommerce/woocommerce-ios/issues/1543

And many more. 

## Changes

The tag is added in `beforeSerializeEvent`:

https://github.com/Automattic/Automattic-Tracks-iOS/blob/9d540115624f6f081ba1dbafc4085b6ff8108116/Automattic-Tracks-iOS/Crash%20Logging/CrashLogging.swift#L67-L75

I created a new class, [`ApplicationFacade`](https://github.com/Automattic/Automattic-Tracks-iOS/blob/add/application-state-tag/Automattic-Tracks-iOS/Crash%20Logging/ApplicationFacade.swift), to hide the complexity of the conditional compilation. I initially added this in `CrashLogging` but I concluded that it was making the class difficult to read (and unnecessarily long). 

## Testing

You can use the WC branch from this PR https://github.com/woocommerce/woocommerce-ios/pull/2377 which is already set up. 

I used the WooCommerce Alpha scheme to test this.

1. Add this line in `viewDidLoad` of `HelpAndSupportViewController` or any part of the code that you fancy. 

    ```swift
            CrashLogging.logError(NSError(domain: "this is a test", code: 5_000, userInfo: nil))
    ```

2. Navigate to Login → Help if you're not logged in. Or Settings → Help & Support if you're logged in. 
3. Wait for the event to arrive in Sentry.
4. Confirm that there is an `app.state` tag added to Sentry.

### About Background State

I found it a bit tricky to test the `app.state` to be “background”. Whenever I debug Xcode, it seems to stop running the code when I place the app in the background. That's expected I guess? The screenshot above showed that I was able to make it log while in the background. I did that by running using this: 

```swift
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(3)) {
            CrashLogging.logError(NSError(domain: "this is a test", code: 5_000, userInfo: nil))
        }
```

Still, it was tricky and seemed to require precise timing. 

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [ ] Create a version tag (?) if approved and merged.
